guava/src/com/google/common/escape/UnicodeEscaper.java

```diff
@@ -127,14 +133,17 @@ public String escape(String string) {
    * @throws IllegalArgumentException if the scanned sub-sequence of {@code csq} contains invalid
    *     surrogate pairs
    */
-  protected int nextEscapeIndex(CharSequence csq, int start, int end) {
-    int index = start;
+  @SuppressWarnings("upperbound:compound.assignment.type.incompatible")/* (1) `Character.isSupplementaryCodePoint(cp)` is true when `cp` range from 65536 to 1114111.
```

Comment:
I don't think this explanation is sufficiently clear. Looking [here](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#unicode), I found what I believe to be the correct explanation: in `char` arrays and `CharSequence` and its subclasses, *two* `char`s are used to represent unicode characters that are outside the range of a 16-bit `char` character. So it's not that `cp` is necessarily in 0-65535 that this code is correct: if that were so, this line would be unnecessary. Rather, it's that if `cp` is in the larger range then there is guaranteed to be another character following the one located at `index`, which is the rest of the unicode character. I think this explanation applies to all of the comments similar to this one.